[WGSL] Add typing for field access expressions

tadeuzagallo opened the pull request:
#### 7b9d9eb03014969d031e19f223e334133081f893
<pre>
[WGSL] Add typing for field access expressions
<a href="https://bugs.webkit.org/show_bug.cgi?id=251842">https://bugs.webkit.org/show_bug.cgi?id=251842</a>
rdar://105120646

Reviewed by Myles C. Maxfield.

Now that we can infer struct types, add support for access struct members. Support
for accessing vector members will follow in a separate patch.

* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::check):
(WGSL::TypeChecker::visitStructMembers):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::inferred):

Canonical link: <a href="https://commits.webkit.org/260196@main">https://commits.webkit.org/260196@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/977726bc2694b427f24ad52cd2b6d8bf0b1348bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107490 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16552 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40393 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116653 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116074 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17977 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7794 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99654 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113251 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13576 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96754 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41197 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95481 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28380 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82971 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9558 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29733 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10215 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6634 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15714 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49314 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7051 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11775 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum reviewed (approved):
Right, super straightforward.

djg commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -205,9 +222,25 @@ void TypeChecker::visit(AST::Expression&)
+    }
+
+    if (std::holds_alternative<Struct>(*baseType)) {
+        auto& structType = std::get<Struct>(*baseType);
```
nit: This feels like the pattern in `WGSL::AST` that is written:
```
    if (is<Struct>(*baseType)) {
        auto& structType = downcast<Struct>(*baseType);
```

I would like to err on the side of consistency in the compiler. (Also, after looking at built times with `-ftime-trace`, I hate how slow `std::variant`s are to compile)

tadeuzagallo replied:
Sure, I can refactor to match that in a separate commit. (Just because it's already used in several other places in the type checker)

tadeuzagallo committed:
[WGSL] Add typing for field access expressions
https://bugs.webkit.org/show_bug.cgi?id=251842
rdar://105120646

Reviewed by Myles C. Maxfield.

Now that we can infer struct types, add support for access struct members. Support
for accessing vector members will follow in a separate patch.

* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::check):
(WGSL::TypeChecker::visitStructMembers):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::inferred):

Canonical link: https://commits.webkit.org/260196@main
